### PR TITLE
Allows API URL override via environment variable to support GitHub Enterprise

### LINKED ISSUE #23: Provide ability to support GitHub Enterprise
The URL defaults to api.github.com in octokit/auth-app.js, which makes it impossible for this action to work with a GHE installation.

### PR DESCRIPTION
**Pull request by retiman**
Fixes #23.

### CODE REVIEW

**Comment by retiman** [comment]
I bumped the version to 1.4.0 to avoid potentially breaking existing users.

**Comment by peter-evans** [comment]
There is a much simpler way to support GHE in actions. I've done the following in my [own actions](https://github.com/peter-evans/create-pull-request/blob/main/src/utils.ts#L42).
```
const githubUrl = process.env['GITHUB_SERVER_URL'] || 'https://github.com'
```

This is also how it's done in the official actions like checkout:
https://github.com/actions/checkout/blob/main/src/url-helper.ts#L22-L29

**Comment by retiman** [comment]
Thanks for the advice Peter.  I can't get this to work on my end, though.  I see your action is mainly working with the actions toolkit (actions/core in the case you linked); however, this action is leveraging octokit directly.  The octokit/auth-app.js library delegates to octokit/endpoint.js, which hard codes `api.github.com` as the default `baseUrl`:

https://github.com/octokit/endpoint.js/blob/master/src/defaults.ts#L12

I don't see any place where auth-app.js, request.js, or endpoint.js are reading `GITHUB_SERVER_URL` or `GITHUB_URL` to override this value.  Do you have any examples where `github-app-token` is generating a token with the appropriate environment variables set?

**Comment by peter-evans** [comment]
For overriding `api.github.com` to support GHE you can do this:

```typescript
    const options: OctokitOptions = {}
    if (token) {
      options.auth = `${token}`
    }
    options.baseUrl = process.env['GITHUB_API_URL'] || 'https://api.github.com'
    this.octokit = new Octokit(options)
```
See: https://github.com/peter-evans/create-pull-request/blob/main/src/github-helper.ts#L27

**Comment by retiman** [comment]
Sure.  That simplifies things.

File: README.md [hunk 1]
```diff
@@ -6,6 +6,8 @@ This [JavaScript GitHub Action](https://help.github.com/en/actions/building-acti
 A workaround is to use a [personal access token](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line) from a [personal user/bot account](https://help.github.com/en/github/getting-started-with-github/types-of-github-accounts#personal-user-accounts).
 However, for organizations, GitHub Apps are [a more appropriate automation solution](https://developer.github.com/apps/differences-between-apps/#machine-vs-bot-accounts).
 
+If using GitHub Enterprise, set the `GITHUB_API_URL` environment variable to override the default API URL.
```
Comment (peter-evans) [inline README.md:9]: I don't think a user would need to set this manually when calling the action. I think this is just a default environment variable inside GHE.

https://docs.github.com/en/enterprise-server@3.1/actions/reference/environment-variables

> GITHUB_API_URL | Returns the API URL. For example: http(s)://[hostname]/api/v3.

I don't run GHE so I can't test this to make 100% sure.

Comment (retiman) [reply README.md:9]: I just checked, yup, it does output the right value.  Removing the changes on README.md.

**Comment by peter-evans** [comment]
Nice. LGTM but I'm not the maintainer 😄

**Comment by retiman** [comment]
> Nice. LGTM but I'm not the maintainer 😄

Thanks for the help though :)

File: src/fetch-installation-token.ts [hunk 2]
```diff
@@ -12,7 +13,13 @@ export const fetchInstallationToken = async ({
   privateKey: string;
   repo: string;
 }>): Promise<string> => {
-  const app = createAppAuth({ appId, privateKey });
+  const app = createAppAuth({
+    appId,
+    privateKey,
+    request: request.defaults({
+      baseUrl: process.env["GITHUB_API_URL"] || "http://api.github.com"
```
Comment (tibdex) [inline src/fetch-installation-token.ts:20]: I think the fallback is unnecessary:

```suggestion
      // GITHUB_API_URL is part of GitHub Actions' built-in environment variables.
      // See https://docs.github.com/en/actions/reference/environment-variables#default-environment-variables.
      baseUrl: process.env["GITHUB_API_URL"]
```

Comment (peter-evans) [reply src/fetch-installation-token.ts:20]: The official toolkit uses the fallback, but I'm not sure if there is a good reason.
https://github.com/actions/toolkit/blob/main/packages/github/src/context.ts#L49

Perhaps it supports self-hosted runners? 🤷‍♂️

**Comment by peter-evans** [comment]
After looking at [getOctokit](https://github.com/actions/toolkit/blob/f0b00fd201c7ddf14e1572a10d5fb4577c4bd6a2/packages/github/src/github.ts) in the github package I'm not sure if this change is necessary. The toolkit already includes this env var lookup for the base URL here: https://github.com/actions/toolkit/blob/f0b00fd201c7ddf14e1572a10d5fb4577c4bd6a2/packages/github/src/internal/utils.ts#L23-L25

@retiman Are you sure this action doesn't work on GHE already?

**Comment by retiman** [comment]
> After looking at [getOctokit](https://github.com/actions/toolkit/blob/f0b00fd201c7ddf14e1572a10d5fb4577c4bd6a2/packages/github/src/github.ts) in the github package I'm not sure if this change is necessary. The toolkit already includes this env var lookup for the base URL here: https://github.com/actions/toolkit/blob/f0b00fd201c7ddf14e1572a10d5fb4577c4bd6a2/packages/github/src/internal/utils.ts#L23-L25
> 
> @retiman Are you sure this action doesn't work on GHE already?

Yea, I verified it doesn't work if `GITHUB_API_URL` is present (verified last night based on your suggestion).  If that variable is leveraged by octokit things would "just work".  

The link you sent me is for the actions toolkit, where that makes more sense.  The toolkit code is meant to be run in build agents and the `GITHUB_API_URL` is expected to be present.  Octokit is a general use library, and its consumers are not necessarily running in a GHA runner so there's no guarantee that the var would be present.

File: src/fetch-installation-token.ts (hunk 2 above)
Comment (retiman) [reply src/fetch-installation-token.ts]: We are using self-hosted runners and that variable is present.  I believe it's added in the runner software so it's probably defensive programming.

**Comment by peter-evans** [comment]
@retiman Let me try and break down the route this action takes to build it's Octokit instance and you'll see that `GITHUB_API_URL` is already being handled.

`getOctokit` is called from the `@actions/github` package of the actions toolkit:
https://github.com/tibdex/github-app-token/blob/main/src/fetch-installation-token.ts#L17

A new `GitHub` is created from `.utils`:
https://github.com/actions/toolkit/blob/f0b00fd201c7ddf14e1572a10d5fb4577c4bd6a2/packages/github/src/github.ts#L19

`GitHub` is a new Octokit instance with `defaults` set:
https://github.com/actions/toolkit/blob/f0b00fd201c7ddf14e1572a10d5fb4577c4bd6a2/packages/github/src/utils.ts#L20-L23

`defaults` includes `baseUrl`, which requires a call to `Utils.getApiBaseUrl()`:
https://github.com/actions/toolkit/blob/f0b00fd201c7ddf14e1572a10d5fb4577c4bd6a2/packages/github/src/utils.ts#L12-L18

`getApiBaseUrl()` handles `GITHUB_API_URL` with a fallback:
https://github.com/actions/toolkit/blob/f0b00fd201c7ddf14e1572a10d5fb4577c4bd6a2/packages/github/src/internal/utils.ts#L23-L25

You can see the code exists in the `dist` minified release of the action:
image

So the Octokit instance used by this action should already have the correct base URL set if `GITHUB_API_URL` is available in the environment that it's running in.

**Comment by retiman** [comment]
@peter-evans I don't deny this is happening at the actions toolkit level; all the links you gave are code paths into `actions/toolkit`.

However, when using a non-modified version of this action in GHE, I'm hitting the 404 error on line 15 in `fetch-installation-token.ts`, before it even its `getOctokit`:

https://github.com/tibdex/github-app-token/blob/main/src/fetch-installation-token.ts#L15

The `createAppAuth` function in `octokit/auth-app.js` calls `octokit/request.js`:

https://github.com/octokit/auth-app.js/blob/master/src/index.ts#L49
https://github.com/octokit/request.js/blob/master/src/index.ts#L7
https://github.com/octokit/request.js/blob/master/src/with-defaults.ts#L15

You'll see in `octokit/endpoint.js` (which is referenced from `with-defaults.ts`), the URL is hard coded:

https://github.com/octokit/endpoint.js/blob/master/src/index.ts#L4
https://github.com/octokit/endpoint.js/blob/master/src/defaults.ts

Stepping back a bit, do you believe that the env var should be checked in octokit?

**Comment by peter-evans** [comment]
Ah, sorry. 🙏    I was getting confused because I didn't read the code properly and assumed this was about the octokit instance in use, but it's about the `createAppAuth` call. So in that case your change is necessary. LGTM!

(No I don't think `GITHUB_API_URL` should be checked in Octokit. As you rightly said, it's a general purpose lib.)

**Comment by peter-evans** [comment]
homer

**Comment by retiman** [comment]
> Ah, sorry. 🙏 I was getting confused because I didn't read the code properly and assumed this was about the octokit instance in use, but it's about the `createAppAuth` call. So in that case your change is necessary. LGTM!
> 
> (No I don't think `GITHUB_API_URL` should be checked in Octokit. As you rightly said, it's a general purpose lib.)

No worries, thanks for following up.

**Review by tibdex — APPROVED** [review]
Thanks, let's see how this works.

**Comment by retiman** [comment]
> Thanks, let's see how this works.

Just tested against GHE; works great.

**Pull request merged**